webtop/api.py

```diff
@@ -0,0 +1,159 @@
+
+
+class Resolver(aiohttp.resolver.AbstractResolver):
+    async def close(self) -> None:
+        await self.async_resolver.close()
+
+    def __init__(self, *args, custom_mappings: Optional[Dict[str, str]] = None, **kwargs):
+        super().__init__(*args, **kwargs)  # type: ignore
+        if custom_mappings is None:
+            self.custom_mappings: Dict[str, str] = {}
+        else:
+            self.custom_mappings = custom_mappings
+        self.async_resolver = aiohttp.resolver.AsyncResolver()  # type: ignore
+
+    async def resolve(self, host: str, port: int = 0, family: int = socket.AF_INET) -> List[Dict[str, Any]]:
+        if host in self.custom_mappings:
+            return [
+                {
+                    "hostname": host,
+                    "host": self.custom_mappings[host],
+                    "port": port,
+                    "family": family,
+                    "proto": 0,
+                    "flags": socket.AI_NUMERICHOST,
+                }
+            ]
+        return await self.async_resolver.resolve(host, port, family)
+
+
+class HTTPMethod(enum.Enum):
+    GET = "GET"
+    HEAD = "HEAD"
+
+
+class Runner(object):
+    def __init__(
+        self,
+        url: str,
+        name_resolution_overrides: Optional[Dict[str, str]],
+        method: str = "GET",
+        number_of_running_requests: int = 100_000,
+        number_of_workers: int = 1,
+        timeout: int = 10,
+    ):
+        self.url = url
+        self.method = HTTPMethod[method]
+        self.__name_resolution_overrides = name_resolution_overrides or {}
+        self.__number_of_running_requests = number_of_running_requests
+        self.__number_of_workers = number_of_workers
+        self.__results: Deque[Result] = deque(maxlen=number_of_running_requests)
+        self.__stop_event = Event()
+        self.__timeout = timeout
+
+    def __session(self) -> aiohttp.ClientSession:
+        return aiohttp.ClientSession(
+            timeout=aiohttp.ClientTimeout(connect=self.__timeout),
+            connector=aiohttp.TCPConnector(
+                force_close=True, limit=0, resolver=Resolver(custom_mappings=self.__name_resolution_overrides)
+            ),
+        )
+
+    async def start(self,) -> None:
+        async with self.__session() as session:
+
+            async def worker() -> None:
+                while not self.__stop_event.is_set():
+                    result = await request(url=self.url, method=self.method.value, session=session)
+                    self.__results.append(result)
+
+            event_loop = asyncio.get_event_loop()
+            tasks = []
+            for _ in range(self.__number_of_workers):
+                tasks.append(event_loop.create_task(worker()))
+            self.__tasks = tasks
+
+    async def stop(self) -> None:
+        self.__stop_event.set()
+        asyncio.gather(*self.__tasks)
+
+    def get_results(self) -> Set[Result]:
+        return set(copy.deepcopy(self.__results))
```

Comment:
Does this need to be a deep copy? There are pickling errors, and everything worked fine if I just removed `copy.deepcopy`.